### PR TITLE
Clipped data

### PR DESCRIPTION
**Pull request by danlipsa**
Fixes for issue #1720

### CODE REVIEW

**Comment by aashish24** [comment]
Nice @danlipsa!

File: Packages/vcs/Lib/VTKPlots.py [hunk 1]
```diff
@@ -1231,9 +1231,9 @@ def fitToViewport(self, Actor, vp, wc=None, geo=None, priority=None,
                 NGridCover = 0
                 pt.SetNumberOfPoints(Npts * Npts)
                 for x in numpy.arange(
-                        Xrg[0], Xrg[1], (Xrg[1] - Xrg[0]) / Npts):
+                        Xrg[0], Xrg[1] + (Xrg[1] - Xrg[0]) / Npts, (Xrg[1] - Xrg[0]) / Npts):
```
Comment (aashish24) [inline Packages/vcs/Lib/VTKPlots.py]: @danlipsa can you explain what's going on here?

**Comment by aashish24** [comment]
@danlipsa you would have to update all the baselines as well I believe. The workflow is that you upload your baselines first and then push the code-branch. since then the buildbots can pick up the right data branch corresponding to the code-branch. Hope this helps.

File: Packages/vcs/Lib/VTKPlots.py (hunk 1 above)
Comment (danlipsa) [inline Packages/vcs/Lib/VTKPlots.py]: Here we generate a 2D grid of points to cover the whole domain and transform those points through the geographic projection. The goal is to get the bounds of the projection. The problem is that arrange uses an interval open at the higher end, so the points on the margin are not included. What I have done is to include the points on the margin.

Of course a better fix would be to pass the bounds of the projection but that will take more time to learn the code.

**Comment by aashish24** [comment]
thanks so if Xarg[1] say is 100, Xarg[0] = 60 and N=50

Then Step = 40/50 = 0? in numpy.arrange? and also 

the arguments to numpy.arrange would be:
(60, 100, 0) Is that correct? which would be open-ended [65, 100)

- Aashish

**Comment by danlipsa** [comment]
I think those are floats so the code uses the correct step.

**Comment by aashish24** [comment]
> I think those are floats so the code uses the correct step.

Ah.. I didn't see that. Thanks, yes, then your change make sense to me and looks correct. Are you going to update the baselines as well?

**Comment by aashish24** [comment]
Ref: https://github.com/UV-CDAT/uvcdat/issues/1725

**Comment by danlipsa** [comment]
@aashish24 yes I wll. I will also add a test for the problem I am solving. I had some issues with a couple of tests producing empty values, so in that case I reverted to what we were doing before. Strangely that was aeqd actually working correctly. I have to see what was the difference between that test and the test that does not work.

**Comment by aashish24** [comment]
@danlipsa the test images are looking much better and correct and that's why they are not matching up with baselines. I would suggest you update the baselines as well. I don't know why garant is not showing these test failures.

**Comment by doutriaux1** [comment]
@danlipsa see: https://open.cdash.org/viewTest.php?onlyfailed&buildid=4138094 newly generated pix seem better than baseline, I would recommend updating.

**Comment by danlipsa** [comment]
@doutriaux1 We found two more projections that needed more points with the new changes. Those were robinson and  mollweide only for meshfill. Do you remember why we compute the bounds of the geographic projection instead of passing it as a parameter?

**Comment by danlipsa** [comment]
@doutriaux1 @aashish24 I would get rid of that computation altogether but probably it is too involved to do it before the release.

**Comment by doutriaux1** [comment]
@danlipsa you're right we should pass it as a parameter, `projection_type:array` in case the projection type changes between refresh. Let's take a look post 2.4

**Comment by aashish24** [comment]
@danlipsa did you update all the baselines as we discussed today?

**Comment by aashish24** [comment]
@williams13 @doutriaux1 we found few issues in the code, hoping that this branch will fix some of them. For others we need to discuss.

**Comment by danlipsa** [comment]
@aashish24 @doutriaux1 I did. It seems that at least for the first file this is a difference between PNG generated on my machine and PNGs generated with MESA. It seems the difference is the black line at the top. Do you guys know what that is?

**Comment by danlipsa** [comment]
@aashish24 @doutriaux1 Note the Mesa generated file has the black line at the bottom. Should we increase the threshold?

**Comment by doutriaux1** [comment]
it's a bit disturbing, the black line is in a layer higher than the data, it should be on top of it. VTK bug? MESA bug?

**Comment by aashish24** [comment]
@danlipsa if you are talking about test_polar_set_opt_param_polar.png then I would say that in the old baseline the continents are clipped on top bottom left and right. So the new baseline seems correct to me.

**Comment by danlipsa** [comment]
@doutriaux1 @aashish24 I think it is fairly irrelevant for the correctness of the results shown if the black line is there or not. I would suggest adding another baseline for these tests. My guess is that there is a difference in the OpenGL drivers for Mesa. I would advise adding another baseline. Note that the tests on   cont-int/LLNL/Darwin-Mac (NOGUI) have all passed. We can also add an issue for someone to look into this after the release, but I don't think we should delay the release for this. Thoughts?

**Comment by danlipsa** [comment]
@aashish24 We are talking about the test results on Mesa
https://open.cdash.org/viewTest.php?onlyfailed&buildid=4138409

**Comment by aashish24** [comment]
thanks @danlipsa  @doutriaux1 are we drawing these boundaries? 

@danlipsa these lines were in the old baseline. If you upload the new baselines, your tests should pass?

**Comment by danlipsa** [comment]
@aashish24 I think the lines were not in the old baselines. I think we do draw them. Maybe the layer with the lines is close enough to the data that Mesa draws is under the data. Note that both my machine and the tests in cont-int/LLNL/Darwin-Mac (NOGUI) shows the lines (these have hardware OpenGL). We did change slightly the parallel projection and probably that made the difference.

**Comment by danlipsa** [comment]
@doutriaux1 Are you saying that the black line should be there? This is what I got with OpenGL on my machine.

**Comment by doutriaux1** [comment]
@aashish24 @danlipsa fyi the competition using proj4 better than us.... http://basemaptutorial.readthedocs.org/en/latest/projections.html

**Comment by aashish24** [comment]
@danlipsa changes looks good to me. 

@doutriaux1 yes, we are going to use latest proj4 and we can use our own. I have a small fix for VTK for that. The only thing we need is the mangled proj4 for VTK.

**Comment by aashish24** [comment]
@doutriaux1 can we merge this branch now?

**Comment by doutriaux1** [comment]
@aashish24 let me run the test suite one last time before merging.

**Comment by aashish24** [comment]
sure @doutriaux1 but what are you concerned about? all tests are passing on garant and mac and the code looks reasonable.

**Comment by doutriaux1** [comment]
just want to add the extra step. There's a lot of yellow on the dashboard.

**Comment by aashish24** [comment]
@doutriaux1 are you talking about buildbots not running?

**Comment by doutriaux1** [comment]
yes @aashish24 mac are slowly catching up, RH need to restart (I really need to figure out which test makes them hang!)

**Comment by aashish24** [comment]
@doutriaux1 any luck?

**Comment by doutriaux1** [comment]
still need to run ctest connection keeps dropping here....

**Comment by aashish24** [comment]
Are you running ctest locally? or are you trying to start the Linux buildbots?

**Comment by doutriaux1** [comment]
locally on my machine at the lab, which means remote X forwarding which is flaky here.

**Comment by aashish24** [comment]
Okay... are you concerened that dashboards are not catching something (garant and mac)..

**Comment by doutriaux1** [comment]
just want to be sure nothing showed up on Lniux, will probably try to retrigger the bots as well.

**Comment by aashish24** [comment]
Oh okay. .. garant seems to be happy. Let me know and I can run tests on my machine as well (in case you are having issues).

**Comment by doutriaux1** [comment]
@aashish24 it was good to wait on the bots. @danlipsa can you please up a bit the thresholds

**Comment by danlipsa** [comment]
@doutriaux1 Do you have the files added by this commit in your mesa buildbot?
https://github.com/UV-CDAT/uvcdat-testdata/commit/d1ef16da34b68a58641ec327b3bc62d5a225a40e
It seems that the errors we get are the same as before I added these files.

**Comment by aashish24** [comment]
@doutriaux1 looks like your bot are not picking the _1 files. Can you check if these files are downloaded on the failing bots?

**Comment by aashish24** [comment]
@danlipsa I updated the branch with current master to see if LLNL bots will pick up update _1 images. If not we would have to work on the threshold for now as @doutriaux1 suggested.

**Comment by doutriaux1** [comment]
@danlipsa do you still need me to merge this or should I wait for #1736 ?

**Comment by doutriaux1** [comment]
@danlipsa it's all good on my ubuntu here.

**Comment by danlipsa** [comment]
@doutriaux1 It would be great if you could merge this. I can rebase #1736 after that.

**Comment by doutriaux1** [comment]
ok so what about the bots? Why are they failing can we up the thresholds?

**Comment by aashish24** [comment]
> ok so what about the bots? Why are they failing can we up the thresholds?

@doutriaux1 I guess you didn't read all the messages. On MESA we are seeing this extra line. I believe it could be because of the polygon offset not working or set on MESA. To resovle this issue for now @danlipsa uploaded _1 versions but seems to me that they are not getting picked up by bots at LLNL. Can you chekc if the data checkout has _1 images? @danlipsa did you try what we discussed this morning? Other thing @danlipsa we could do is set the global offset in VTK to have the lines always show up on top?

**Comment by doutriaux1** [comment]
@aashish24 I did read the thread and I did reply the line `should` always be on top! It's the same priority as the data and drawn later. So that needs to be addressed. I will check to see if the _1 are picked on bots.

**Comment by doutriaux1** [comment]
@aashish24 @danlipsa we can solve that line issue post 2.4. But we need the bots to be reasonably ok before the release. Will check on the _1 on bots.

**Comment by aashish24** [comment]
oh okay thanks @doutriaux1. @danlipsa try setting the polygonoffset? (or lineoffset?)

**Comment by danlipsa** [comment]
This has new baselines
https://github.com/UV-CDAT/uvcdat-testdata/pull/92

**Comment by danlipsa** [comment]
@doutriaux1 @aashish24 I tried to switch the regular png with the _1 png for one of the tests to make sure ctest works correctly. It does. The test still passed on my machine even if the png that matched was the _1 one.

**Comment by danlipsa** [comment]
@doutriaux1 @aashish24 So, it does appear that maybe the buildbots do not pull the _1 files correctly.

**Comment by doutriaux1** [comment]
@danlipsa I just double check after a retrigger the _1 files are here for oceanonly ( http://aims4.llnl.gov:9981/master/macbot2.llnl.gov:9982/UV-CDAT/uvcdat/8055818c0a23e7019b638dbb2ed7e49eb95a7999/ctest__-j4__-D )

I'm not sure what you're trying to do ,but the files are defintely there, what are the _1 for?

The way the process work on the bot is that master is checked out then the branch `clipped-data` then master is merged back in, would that throw off your test?

**Comment by doutriaux1** [comment]
runnnig the ctest on your branch with uvcdat-data having been subjected to a:  `git reset --hard origin/clipped-data`. Will post results soon.

**Comment by doutriaux1** [comment]
@aashish24 @danlipsa still fails all the test, will post the link when done. What are the _1 for?

**Comment by doutriaux1** [comment]
Could it be due changes in master that are put back into your branch? Can you trst on your computer after you merged master back into `clipped_data`?

**Comment by doutriaux1** [comment]
okay failing link: https://open.cdash.org/viewTest.php?onlyfailed&buildid=4148428

**Comment by danlipsa** [comment]
@doutriaux1 The _1 files are alternate baselines. ctest looks for alternate baselines if the main image fails the test.

**Comment by danlipsa** [comment]
@doutriaux1 @aashish24 I could recompile VTK with Mesa on my machine and try that. Hopefully it will have the same behavior as the test machine.

**Comment by aashish24** [comment]
@danlipsa why don't we set line offset so that lines are always drawn on top. May be it will take care of some of the issues. I don't think it will take care of all issues..

**Comment by danlipsa** [comment]
@doutriaux1 Can you try running the first test that fails from command line:vcs_test_basic_meshfill_aeqd_proj_SH_gm Does it fail?
If yes, can you try copying the _1 baseline over the regular baseline:
cp test_vcs_basic_meshfill_aeqd_proj_SH_via_gm_1.png test_vcs_basic_meshfill_aeqd_proj_SH_via_gm.png

Run the test again. Does it fail?
This will give us more information about what is wrong with the test machine or with the test.

**Comment by danlipsa** [comment]
@aashish24 I will try looking at this maybe I see something. The problem is that I don't have the machine where the line is not drawn so I have no way to test.

**Comment by danlipsa** [comment]
@aashish24 @doutriaux1 I'll use OSMesa maybe this will have the same behavior as the test machine.

**Comment by danlipsa** [comment]
@aashish24 @doutriaux1 I compiled with OSMesa and indeed it does not seem to pickup the _1 file. I will look into why.

**Comment by doutriaux1** [comment]
@danlipsa @aashish24 I'm still perplexed by this `_1` business. I'd say increase the threshold for the release and let's try to fix it properly post-release. Thoughts?Although MESA seems to be the only one that does things correctly with the line on top. So it may be a VTK or opengl issue.

**Comment by danlipsa** [comment]
@doutriaux1 @aashish24 The reason why the test would not pickup the _1 is a parameter to checkimage.check_result_image that said it should not. See my latest commit. Hopefully this will make all tests pass. I can keep looking into why the line does not show up on OSMesa but if all the tests pass we can release before that.

**Comment by danlipsa** [comment]
@doutriaux1 @aashish24 Ready to merge? 5/9 are green and it has been running for a while. Not sure if anything else is going to finish.

**Comment by danlipsa** [comment]
@aashish24 @doutriaux1 Looking at why on OSMesa we don't have line showing, it seems that this is a fitToViewport issue as well rather then a polygonoffset/lineoffset issue. Given that the parallel projection depends on the matrix of dots we generate it makes sense that this problem showed up when I changed the matrix of points in this bug fix. My opinion is that we should look into passing the actual bounds of the projected dataset. After that we will see if we still have the line problem on OSMesa. Thoughts?

**Comment by aashish24** [comment]
> as well rather then a polygonoffset/lineoffset issue. Given that the parallel projection depends on the matrix of dots we generate it makes sense that this problem showed up when I changed the matrix of points in this bug fix. My opinion is that we should look into passing the actual bounds of the projected dataset. After that we will see if we still have the line problem on OSMesa. Thoughts?

+1 can you file a bug for this @danlipsa 

@danlipsa the dashboards are looking pretty good now.. failures on garant seems to be an issue related to not pulling in the right test data branch. May be you need to update your data branch as well. But other than that.. its +1 for me. @doutriaux1 ?

**Comment by danlipsa** [comment]
@aashish24 See https://github.com/UV-CDAT/uvcdat/issues/1739

**Comment by danlipsa** [comment]
@doutriaux1 Is this ready to merge?

**Comment by aashish24** [comment]
dan lets merge this branch. Any remainig concerns if not found or discussed can be handled in another pr.

**Pull request merged**